botorch/generation/gen.py

```diff
@@ -109,6 +114,16 @@ def f(x):
         loss = -acquisition_function(X_fix).sum()
         # compute gradient w.r.t. the inputs (does not accumulate in leaves)
         gradf = _arrayify(torch.autograd.grad(loss, X)[0].contiguous().view(-1))
+        if np.isnan(gradf).any():
+            raise RuntimeError(
+                f"{np.isnan(gradf).sum()} elements of the {x.shape} gradient tensor "
+                "`gradf` are NaN."
+                + (
+                    "Consider using `dtype=torch.double`."
+                    if initial_conditions.dtype != torch.double
+                    else ""
+                )
```

Comment:
Let's make this a bit more readable

```suggestion
            msg = (
                f"{np.isnan(gradf).sum()} elements of the {x.shape} gradient tensor "
                " are NaN. This often indicates numerical issues."
            )
            if initial_conditions.dtype != torch.double:
                msg += " Consider using `dtype=torch.double`."
            raise RuntimeError(msg)
```